### ISSUE TITLE
Introduce color modifiers

### ISSUE THREAD

**Issue by six7** [tokens-studio/figma-plugin#1166]
## Description
Users often need to manipulate colors to create slight color variations of a specific base color. An example would be a hover color that is just slightly brighter than the other one, or a pressed state which is usually a bit darker. Also users would want to create alpha variations of a color. This often depends on a few things:

## Type of operation
- Lighten: The amount to lighten by
- Darken: The amount to darken by
- Mix: The color you want to mix with and the amount you want this other color to come through
- Alpha: The amount you want to change the alpha to
 
We could introduce this by implementing a color library such as [Chroma.js](https://gka.github.io/chroma.js/), which does have support for all of the manipulations we need to do:
- Lighten: https://gka.github.io/chroma.js/#color-brighten
- Darken: https://gka.github.io/chroma.js/#color-darken
- Mix: https://gka.github.io/chroma.js/#chroma-mix
- Alpha: https://gka.github.io/chroma.js/#color-alpha

Users would need to configure that in the UI, and we should make it as easy as possible. Let's start with these 4 modifiers and later on we'll see what else is required.

![image](https://user-images.githubusercontent.com/4548309/190499643-6bdf946e-f382-4bc8-987a-58d328083418.png)

As we need to capture these decisions in JSON and these would not be matching the official W3C spec for color values, we need to define these in the $extensions property that we can introduce, which [is part of the W3C spec](https://tr.designtokens.org/format/#extensions) and could be a good starting point to have this kind of modification in place.

The JSON could look like this:

```
{
  "blue": {
    "value": "#0f78dd",
    "type": "color"
  },
  "red": {
    "value": "#ff0000",
    "type": "color"
  },
  "primary": {
    "default": {
      "value": "{blue}",
      "type": "color"
    },
    "lighten": {
      "value": "{primary.default}",
      "type": "color",
      "$extensions": {
        "com.figmatokens": {
          "modify": {
            "type": "lighten",
            "amount": 1.5,
            "space": "LCH"
          }
        }
      }
    },
    "darken": {
      "value": "{primary.default}",
      "type": "color",
      "$extensions": {
        "com.figmatokens": {
          "modify": {
            "type": "darken",
            "amount": 1.5,
            "space": "LCH"
          }
        }
      }
    },
    "mix": {
      "value": "{primary.default}",
      "type": "color",
      "$extensions": {
        "com.figmatokens": {
          "modify": {
            "type": "mix",
            "color": "{red}",
            "amount": 0.5,
            "space": "LCH"
          }
        }
      }
    },
    "alpha": {
      "value": "{primary.default}",
      "type": "color",
      "$extensions": {
        "com.figmatokens": {
          "modify": {
            "type": "alpha",
            "amount": 0.25,
            "space": "LCH"
          }
        }
      }
    },
  }
}
```

## Options on each modifier
| Modifier | Options |
| --- | --- |
| Lighten | Amount (number field) |
| Darken | Amount (number field) |
| Mix | Color (color field with picker and token dropdown showing color tokens), Ratio (number field) 
| Alpha | Opacity (number field between 0 and 1) |

## Available color spaces
- LCH
- sRGB
- P3

**Comment by c1rrus** [comment]
Similar ideas have been discussed in this DTCG issue too: https://github.com/design-tokens/community-group/issues/88

Would be great to share your ideas there too. Perhaps the community can align on something together

**Comment by bewarai** [comment]
The problem I have with this is that the darken or lighten in chroma.js (and many programmers follow the same logic) is that the color system HSL used is not very designsystem friendly towards contrast usability and in line with perceptual color space. I use LCH to output my different variations. That way I can tokenize a color red 20 and align it with green 20 with the same contrast ratio. So I can interchange the color without having it an impact on my contrast.

<img width="760" alt="Picture 1" src="https://user-images.githubusercontent.com/33145617/184921283-de484724-532f-40a3-b7ca-41016f959dfc.png">

**Comment by six7** [comment]
> The problem I have with this is that the darken or lighten in chroma.js (and many programmers follow the same logic) is that the color system HSL used is not very designsystem friendly towards contrast usability and in line with perceptual color space. I use LCH to output my different variations. That way I can tokenize a color red 20 and align it with green 20 with the same contrast ratio. So I can interchange the color without having it an impact on my contrast. So at least let us choose which color-space the darken and light is applied to.

Chroma.js would allow us to choose a colorspace: https://gka.github.io/chroma.js/#chroma-mix so adding to these modifiers would definitely make sense 👍

**Comment by bewarai** [comment]
Some links that can help:
https://www.bugsnag.com/blog/chromatic-sass
https://github.com/aotearoan/sass-farbig

**Comment by AndySobol** [comment]
Hurrah! I look forward to it! ❤️

**Comment by EloB** [comment]
I'm not aware if you maybe already started implementing this but Colord looks like a really good alternative to Chroma.js.

Accoding to their docs they have better performance (300%). Also smaller in size and has plugin based approach if you need to extend it. Builtin Typescript support.

https://github.com/omgovich/colord#benchmarks

Apart from this. I'm like shaking of excitement for one. You rock! Go go go! :D

**Comment by yinonov** [comment]
SD transitive-transforms example shares a similar approach at

https://github.com/amzn/style-dictionary/blob/28787befb1a2c7173c64a2a4bef5b029d0799cce/examples/advanced/transitive-transforms/tokens/color/overlay.json5#L13-L21

**Comment by yinonov** [comment]
how would this be interpreted by Figma in practice? say a rectangle, a 2 color mix would output something like
```css
background: linear-gradient(0deg, #0D0D0D, #0D0D0D),
linear-gradient(0deg, rgba(255, 255, 255, 0.12), rgba(255, 255, 255, 0.12));
```
<img width="605" alt="image" src="https://user-images.githubusercontent.com/15140652/186929092-445edc23-96ea-4a29-ba0c-831cd2f47ac2.png">

would the plugin output a computed primitive value?

**Comment by EloB** [comment]
@six7 I'm not sure if this would be easy to implement but there are some working draft in CSS to include color manipulation. I read about it here:
https://www.digitalocean.com/community/tutorials/css-color-function

In that blog post they said that cssnext, a postcss module, already built this.
https://cssnext.github.io/features/#color-function

So you might be able to reuse that.

If this draft getting accepted you will have working "css" out of the box.

**Comment by yinonov** [comment]
@EloB CSS Color Module Level 5 [color-mix](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/color-mix) will enable run-time color manipulation.

scss and the other transpilers you mention can SSRly manipulate colors and output a primitive value to be used in run-time.

there are other workarounds but it's an author's choice on how to handle in development.

Figma is less flexible in that sense and it's more interesting if this can actually work in Figma?

**Comment by six7** [comment]
Updated mocks to include color space as an option (please let me know if a color space is missing, we'll only ship a subset)

**Comment by EloB** [comment]
@yinonov I understand that but you might be able to use the some of code to achieve realtime. :)

**Comment by LiamMartens** [comment]
@six7 does it make sense to make these add these as value functions instead? ie `lighten({colors.red.500}, 50%)`

**Comment by six7** [comment]
That would be awesome! And yeah, very good point on the array. I think storing it as an array to future-proof this makes lots of sense. I could see us combining these in the future, so let's do that!

One thing that might change is our `$extensions` id (com.figmatokens) in case there's going to be a name change in the future.

**Comment by 1pxone** [comment]
Hey @six7

Any updates on this? Can I help with something to make it happen sooner?

**Comment by AndySobol** [comment]
Hi @six7 =) 

Wanted to ask if this feature will be available in the near future or not to wait this year (2022) ? It is important to know for the project. 

Thank you for your efforts!

**Comment by six7** [comment]
@AndySobol we're focusing on this in the next iteration, after we released the current iteration (early December). The next one after that will likely land in January 2023.

**Comment by EloB** [comment]
@six7 This is me right now!
![excited](https://user-images.githubusercontent.com/476567/204135775-b4b3a595-5a09-4c6c-96e0-4ba94c0b7008.gif)

Sending some love your way.
![love](https://user-images.githubusercontent.com/476567/204135787-7ff75844-f739-4078-853f-1b7e20857ad0.gif)

**Comment by yinonov** [comment]
@six7 is it documented anywhere? curious to see your roadmap

**Comment by AndySobol** [comment]
> @AndySobol we're focusing on this in the next iteration, after we released the current iteration (early December). The next one after that will likely land in January 2023.

@six7 Thank you, so much!!) Thanks for your work!

**Comment by AndySobol** [comment]
> is it documented anywhere? curious to see your roadmap

@yinonov Everything is documented here
https://github.com/users/six7/projects/4/views/17

**Comment by AndySobol** [comment]
Hi!

Are you planning to add the BlendMode type? 
[Figma Api BlendMode](https://www.figma.com/plugin-docs/api/BlendMode/)

**Comment by six7** [comment]
This was released in 1.35 🎉

Read more here: https://docs.tokens.studio/available-tokens/color-tokens#modifiers-pro and https://tokens.studio/changelog/release-1-35
